### PR TITLE
[25092] Document Enhancements

### PR DESCRIPTION
**Pull request by andersdd**
Allow document to be assigned to multiple sources. 
Document display for searching/listing documents in the system.
Assign characteristics to documents.  
Document assignment notes.
Rudimentary document privilege system.

### CODE REVIEW

File: widgets/docCluster.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+DocClusterLineEdit::DocClusterLineEdit(QWidget* pParent, const char* pName) :
+    VirtualClusterLineEdit(pParent, "files", "files_id", "files_descrip", "files_title", 0, 0, pName)
+{
+    setTitles(tr("Document"), tr("Documents"));
+    setUiName("docAttach");
+    setEditPriv("ViewDocuments");
+    setViewPriv("ViewDocuments");
+}
```
Comment (gilmoskowitz) [inline widgets/docCluster.cpp]:
```bash
for i in 1 2 3 ; do echo Recursion is my friend ; done
```

File: guiclient/displays/dspDocuments.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                                tr("Error Getting Assignment"),
+                                qfile, __FILE__, __LINE__))
+    return;
+  else
+  {
+    QUrl url(qfile.value("url_url").toString());
+    if (url.scheme().isEmpty())
+      url.setScheme("file");
+    QDesktopServices::openUrl(url);
+    return;
```
[the remark this answers is not in the document]
Comment (andersdd) [reply guiclient/displays/dspDocuments.cpp]: I will add error handling and variable declaration as you suggest but we need to handle the actual `DesktopServices::openUrl()` opening differently due to the DocumentWatch we have to place on opened files.

[the remark this answers is not in the document]
Comment (andersdd) [reply guiclient/displays/dspDocuments.cpp]: Also need to duplicate back on documents.cpp widget where this code was copied from.

File: widgets/docAttach.cpp [hunk 3]
```diff
@@ -284,12 +305,16 @@ void docAttach::set(const ParameterList &pParams)
...
         _url->setText(url.toString());
       }
+       _notes->setText(qry.value("docass_notes").toString());
       _mode = "edit";
       _docType->setEnabled(false);
     }
     ErrorReporter::error(QtCriticalMsg, 0, tr("Error URL"),
                          qry, __FILE__, __LINE__);
   }
+qDebug() << "sourceType" << _sourcetype << "sourceid" << _sourceid << "urlid" << _id;
```
[the remark this answers is not in the document]
Comment (andersdd) [reply widgets/docAttach.cpp]: Woops - forgot to take out by debug statement

File: guiclient/menuCRM.cpp [hunk 4]
```diff
@@ -120,7 +123,11 @@ menuCRM::menuCRM(GUIClient *Pparent) :
     // CRM | Address
     { "menu",		tr("A&ddress"),		(char*)addressMenu,	crmMenu,		"true", NULL, NULL, true	, NULL },
     { "crm.address",	tr("&New..."),		SLOT(sAddress()),	addressMenu,	"MaintainAddresses", NULL, NULL, true	, NULL },
-    { "crm.addresses",	tr("&List..."),	SLOT(sAddresses()),	addressMenu,	"MaintainAddresses ViewAddresses", NULL, NULL, true , NULL },
+    { "crm.addresses",	tr("&List..."),	        SLOT(sAddresses()),	addressMenu,	"MaintainAddresses ViewAddresses", NULL, NULL, true , NULL },
+
+    // CRM | Documents
+    { "menu",		tr("Documents"),        (char*)documentsMenu,	crmMenu,	"true", NULL, NULL, true	, NULL },
+    { "crm.documents",	tr("&List..."),	        SLOT(sDocuments()),	documentsMenu,	"ViewDocuments", NULL, NULL, true , NULL },
```
[the remark this answers is not in the document]
Comment (andersdd) [reply guiclient/menuCRM.cpp]: I sort of agree.  But I also have a configuration setting in CRM. Not sure where it best sits under System-> Setup unless I create a new Configure option (which might give us some future-proofing for future document features)

File: widgets/docAttach.cpp [hunk 5]
```diff
@@ -160,9 +165,21 @@ docAttach::docAttach(QWidget* parent, const char* name, bool modal, Qt::WindowFl
...
                      " WHERE source_widget != ''"
                      " UNION SELECT -2, 'File',     'FILE'"
-                     " UNION SELECT -3, 'Web Site', 'URL') data"
+                     " UNION SELECT -3, 'Web Site', 'URL' "
+                     " UNION SELECT -4, 'Existing Document', 'XFILE') data"
                      " ORDER BY source_descrip;");
 
+  _charassFile->setType("FILE");
+  _charassUrl->setType("FILE");
+  _charid = qrand() % 100000000;
```
[the remark this answers is not in the document]
Comment (andersdd) [reply widgets/docAttach.cpp]: How?? This gives me a random number up to 99,999,999.
I wanted to use `QRandomGenerator` but it is not available till 5.10

File: widgets/docCluster.cpp (hunk 1 above)
Comment (andersdd) [reply widgets/docCluster.cpp]: Don't understand your comment here.

File: widgets/docAttach.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gilmoskowitz) [reply widgets/docAttach.cpp]: Maybe I'm missing something...

The ids of actual documents are generated using a serial column starting with 1. We expect few gaps. As the number of attached files and URLs grows, a random value between 0 and 99 999 999 is increasingly likely to come up with an id already used by an existing file or URL. When there are

this many files/urls | the probability of duplicating an existing id (if there are no gaps)
------ | ----
10 | 1 / 10 ^ 7
1000 | 1 / 10 ^ 5
10 000 | 1 / 10 ^ 4

The insertion of docass records will go swimmingly. The updates, however, will be increasingly likely to pull along pre-existing document associations.

Consider trying the following:
```c++
_charid = qrand(); // about 20x safer
_charid -= _charid % 100 000 000; // removes conflicting ids until count(*) approaches 10^8
```

**Comment by gilmoskowitz** [comment]
OK to merge when xtuple/xtuple#3408 passes CI

**Pull request merged**